File: TPs/Entregables/entrega2/main.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef struct building_t {
    int height;
    int width;
} building_t;

typedef struct stat_t {
    int up;
    int down;
} stat_t;

#define MAX(a, b) ((a) > (b) ? (a) : (b))

void longest_subsequence_with_tail(building_t *buildings, int buildings_size, int index, stat_t *cache) {

    cache[index].up = buildings[index].width;
    cache[index].down = buildings[index].width;

    int max_previous_height = 0;
    int min_previous_height = -1;

    if (index == 0) {
        return;
    }

    int i;
    for(i = index - 1; i >= 0; i--) {
        if (buildings[i].height < buildings[index].height) {
            max_previous_height = buildings[i].height;
            cache[index].up = MAX(cache[index].up, cache[i].up + buildings[index].width);
            continue;
        }

        if (buildings[i].height > buildings[index].height) {
            min_previous_height = buildings[i].height;
            cache[index].down = MAX(cache[index].down, cache[i].down + buildings[index].width);
            continue;
        }
    }
}

stat_t solve(building_t *buildings, int buildings_size) {
    stat_t* cache = malloc(sizeof(stat_t) * buildings_size);

    int i;
    for(i = 0; i < buildings_size; i++) {
        longest_subsequence_with_tail(buildings, buildings_size, i, cache);
    }

    stat_t stats = { .up = 0, .down = 0 };
    for(i = 0; i < buildings_size; i++) {
        stats.up = MAX(stats.up, cache[i].up);
        stats.down = MAX(stats.down, cache[i].down);
    }

    free(cache);
    return stats;
}

int main() {
    int cases;
    scanf("%d", &cases);

    stat_t case_stats[cases];

    int case_num;
    for (case_num = 0; case_num < cases; case_num++) {

        int skyline_len;
        scanf("%d", &skyline_len);

        building_t buildings[skyline_len];
        int aux = 0;

        int i;
        for (i = 0; i < skyline_len; i++) {
            scanf("%d", &aux);
            buildings[i].height = aux;
        }
        for (i = 0; i < skyline_len; i++) {
            scanf("%d", &aux);
            buildings[i].width = aux;
        }

        case_stats[case_num] = solve(buildings, sizeof(buildings) / sizeof(buildings[0]));
    }

    for (case_num = 1; case_num <= cases; case_num++) {
        printf("Case %d. ", case_num);
        if (case_stats[case_num - 1].up >= case_stats[case_num - 1].down) {
            printf("Increasing (%d). Decreasing (%d).\n", case_stats[case_num - 1].up, case_stats[case_num - 1].down);
        }
        else {
            printf("Decreasing (%d). Increasing (%d).\n", case_stats[case_num - 1].down, case_stats[case_num - 1].up);
        }
    }
}

File: TPs/Entregables/entrega2/main.py
class Building:
    def __init__(self, index, height, width):
        self.index = index
        self.height = height
        self.width = width

class Node:
    def __init__(self, building):
        self.building = building
        self.upper_width = 0
        self.lower_width = building.width
        self.childs = None
        self.lowest_child_height = -1

    def insert(self, building):

        if self.building.height >= building.height:
            return

        if self.lowest_child_height >= building.height:
            child = Node(building)
            child.upper_width = self.upper_width + child.building.width
            self.lower_width = max(self.lower_width, self.upper_width + child.building.width)
            self.lowest_child_height = building.height
            self.childs.append(child)
            return

        if self.childs is None:
            child = Node(building)
            child.upper_width = self.upper_width + child.building.width
            self.childs = [child]
            self.lowest_child_height = building.height
            return

        for c in self.childs:
            c.insert(building)

def longest_subsequence_with_tail(buildings, index, cache):

    if index == 0:
        cache = { index : buildings[index][1] }
    else:
        cache[index] = buildings[index][1]

    for k in range(index):
        if buildings[k][0] >= buildings[index][0]:
            continue
        cache[index] = max(cache[k] + buildings[index][1], cache[index])

    return cache

def solve(buildings):
    cache = None

    for k in range(len(buildings)):
        cache = longest_subsequence_with_tail(buildings, k, cache)

    return max(cache.values())

if __name__ == '__main__':
    n_cases = int(input())
    test_cases = []

    for j in range(n_cases):
        buildings = []
        test_cases.append(buildings)

        n_buildings = int(input())

        buildings_heights = input().split()
        buildings_widths = input().split()

        for i in range(n_buildings):
            buildings.append((int(buildings_heights[i]), int(buildings_widths[i])))

    for i in range(n_cases):
        print("Case " + str(i + 1) + ". ", end="")
        inc, dec = solve(test_cases[i])

        if inc >= dec:
            print("Increasing (" + str(inc) + "). ", end="")
            print("Decreasing (" + str(dec) + ").")
        else:
            print("Decreasing (" + str(dec) + "). ", end="")
            print("Increasing (" + str(inc) + ").")


#
#
#def solve2(buildings):
#
#    nodes = Node(Building(-1, -1, 0))
#
#    for k in range(len(buildings)):
#        nodes.insert(buildings[k])
#
#    print(nodes.upper_width)
#    print(nodes.childs[0].building.height)
#    print(nodes.childs[0].childs[0].building.height)
#    print(nodes.childs[0].childs[0].childs[0].building.height)
#    print(nodes.childs[0].childs[0].childs[1].building.height)
#    print(nodes.childs[0].childs[0].childs[0].childs[0].building.height)
#    print(nodes.childs[0].childs[0].childs[1].childs[0].building.height)
#
#buildings = [
#        Building(0, 1, 1),
#        Building(0, 4, 1),
#        Building(0, 10, 1),
#        Building(0, 5, 1),
#        Building(0, 20, 1),
#]
#
#solve2(buildings)

# def solve(buildings):
#     n_buildings = len(buildings)
# 
#     cache_increasing = [-1] * n_buildings
#     cache_decreasing = [-1] * n_buildings
# 
#     for j in range(n_buildings):
#         max_increasing = 0
#         max_decreasing = 0
# 
#         for i in range(n_buildings - 1 - j, n_buildings):
#             if (buildings[i][0] < buildings[n_buildings - 1 - j][0]):
#                 max_decreasing = max_decreasing if cache_decreasing[i] < max_decreasing else cache_decreasing[i]
# 
#             if (buildings[i][0] > buildings[n_buildings - 1 - j][0]):
#                 max_increasing = max_increasing if cache_increasing[i] < max_increasing else cache_increasing[i]
# 
#         cache_increasing[n_buildings - 1 - j] = buildings[n_buildings - 1 - j][1] + max_increasing
#         cache_decreasing[n_buildings - 1 - j] = buildings[n_buildings - 1 - j][1] + max_decreasing
# 
#     return (max(cache_increasing), max(cache_decreasing))
# 
# if __name__ == '__main__':
#     n_cases = int(input())
#     test_cases = []
# 
#     for j in range(n_cases):
#         buildings = []
#         test_cases.append(buildings)
# 
#         n_buildings = int(input())
# 
#         buildings_heights = input().split()
#         buildings_widths = input().split()
# 
#         for i in range(n_buildings):
#             buildings.append((int(buildings_heights[i]), int(buildings_widths[i])))
# 
#     for i in range(n_cases):
#         print("Case " + str(i + 1) + ". ", end="")
#         inc, dec = solve(test_cases[i])
# 
#         if inc >= dec:
#             print("Increasing (" + str(inc) + "). ", end="")
#             print("Decreasing (" + str(dec) + ").")
#         else:
#             print("Decreasing (" + str(dec) + "). ", end="")
#             print("Increasing (" + str(inc) + ").")

